{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load libraries\n",
    "\n",
    "library(GenomicRanges, verbose=FALSE) \n",
    "library(pheatmap) \n",
    "library(dplyr)\n",
    "library(tidyr)\n",
    "library(ggplot2)\n",
    "library(DESeq2)\n",
    "library(colorRamp2)\n",
    "library(ComplexHeatmap)\n",
    "library(tximport)\n",
    "\n",
    "##Creating Graphs\n",
    "### Import raw (gene) count data ###\n",
    "\n",
    "directory = \"/Users/tisesuzuki/RNAseq/htseq\"\n",
    "\n",
    "sampleFiles <- grep(\"PDX4\",list.files(directory),value=TRUE)\n",
    "sampleCondition <- c('PDX4_CR', 'PDX4_CR', 'PDX4_CR',\n",
    "                     'PDX4_SE', 'PDX4_SE', 'PDX4_SE')\n",
    "sampleTable <- data.frame(sampleName = sampleFiles,\n",
    "                          fileName = sampleFiles,\n",
    "                          condition = condition)\n",
    "sampleTable$condition <- factor(sampleTable$condition)\n",
    "\n",
    "ddsHTSeq <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTable,\n",
    "                                       directory = directory,\n",
    "                                       design= ~ condition)\n",
    "ddsHTSeq\n",
    "\n",
    "### Filter out genes with counts of less than 10 in all samples ###\n",
    "\n",
    "keepGenes <- rowSums(counts(ddsHTSeq)) >= 10\n",
    "dds_1 <- ddsHTSeq[keepGenes,]\n",
    "summary(dds_1)\n",
    "dim(dds_1)\n",
    "\n",
    "dds_1$condition <- relevel(dds_1$condition, ref = \"PDX4_SE\")\n",
    "dds_1 <- DESeq(dds_1)\n",
    "vsdata <- vst(dds_1, blind=FALSE)\n",
    "plotPCA(vsdata, intgroup = \"condition\")\n",
    "plotDispEsts(dds_1)\n",
    "res <- results(dds_1, contrast = c(\"condition\", \"PDX4_SE\", \"PDX4_CR\"))\n",
    "res\n",
    "sigs <- na.omit(res)\n",
    "sigs <- sigs[sigs$padj < 0.1,]\n",
    "sigs\n",
    "\n",
    "write.csv(res, file = \"deseq_results.csv\")\n",
    "\n",
    "df <- as.data.frame(sigs)\n",
    "df\n",
    "\n",
    "#getting transcript ID - gene symbol table\n",
    "mart <- biomaRt::useMart(\"ensembl\", dataset=\"hsapiens_gene_ensembl\")\n",
    "t2g <- biomaRt::getBM(attributes = c(\"ensembl_transcript_id\", \"ensembl_gene_id\",\n",
    "                                     \"external_gene_name\"), mart = mart)\n",
    "t2g <- dplyr::rename(t2g, target_id = ensembl_transcript_id,\n",
    "                     ens_gene = ensembl_gene_id, ext_gene = external_gene_name)\n",
    "\n",
    "genes <- unique(t2g[,2:3]) #select unique gene names\n",
    "\n",
    "keys <- genes$ens_gene\n",
    "values <- genes$ext_gene\n",
    "l <- list()\n",
    "for (i in 1:length(keys)){\n",
    "  l[keys[i]] <- values[i]\n",
    "}\n",
    "#for non-mapped labels\n",
    "no_values <- setdiff(rownames(df), keys)\n",
    "for (i in 1:length(no_values)){\n",
    "  l[no_values[i]] <- 'NA'\n",
    "}\n",
    "df$symbol <- unlist(l[rownames(df)], use.names = FALSE)\n",
    "\n",
    "df_top <- df[df$padj <= 0.1 & (abs(df$log2FoldChange) > 0.58),]\n",
    "df_top <- df_top[order(df_top$log2FoldChange, decreasing = TRUE),]\n",
    "\n",
    "rlog_out <- rlog(dds_1, blind = FALSE)\n",
    "\n",
    "sampleTable_2 <- sampleTable[,-1]\n",
    "rownames(sampleTable_2) <- sampleTable[,1]\n",
    "coldata <- sampleTable_2[-c(1)]\n",
    "mat <- assay(rlog_out)[rownames(df_top), rownames(coldata)]\n",
    "colnames(mat) <- rownames(coldata)\n",
    "base_mean <- rowMeans(mat)\n",
    "mat.scaled <- t(apply(mat,1,scale))\n",
    "colnames(mat.scaled) <- colnames(mat)\n",
    "\n",
    "num_keep <- 25\n",
    "rows_keep <- c(seq(1:num_keep), seq((nrow(mat.scaled)-num_keep), nrow(mat.scaled)))\n",
    "\n",
    "l2_val <- as.matrix(df_top[rows_keep,]$log2FoldChange)\n",
    "colnames(l2_val) <- \"logFC\"\n",
    "\n",
    "mean <- as.matrix(df_top[rows_keep,]$baseMean)\n",
    "colnames(mean) <- \"AvgExp\"\n",
    "\n",
    "col_logFC <- colorRamp2(c(min(l2_val),0, max(l2_val)), c(\"gray50\", \"white\", \"#BE575F\"))\n",
    "col_AvgExp <- colorRamp2(c(quantile(mean)[1], quantile(mean)[4]), c(\"white\", \"#941A2B\"))\n",
    "\n",
    "ha <- HeatmapAnnotation(summary = anno_summary(gp = gpar(fill = 2), \n",
    "                                               height = unit(2, \"cm\")))\n",
    "\n",
    "h1 <- Heatmap(mat.scaled[rows_keep,], cluster_rows = F, colorRampPalette(c(\"black\",\"white\",\"#941A2B\"))(50),\n",
    "              column_labels = colnames(mat.scaled), name=\"Z-score\",\n",
    "              cluster_columns = T)\n",
    "h2 <- Heatmap(l2_val, row_labels = df_top$symbol[rows_keep], \n",
    "              cluster_rows = F, name=\"logFC\", top_annotation = ha, col = col_logFC,\n",
    "              cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid\n",
    "                grid.text(round(l2_val[i, j],2), x, y)\n",
    "              })\n",
    "h3 <- Heatmap(mean, row_labels = df_top$symbol[rows_keep], \n",
    "              cluster_rows = F, name = \"AvgExp\", col=col_AvgExp,\n",
    "              cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid\n",
    "                grid.text(round(mean[i, j],2), x, y)\n",
    "              })\n",
    "\n",
    "h<-h1+h2+h3\n",
    "h\n",
    "png(\"~/RNAseq/results/20240108_heatmap.png\", res = 300, width = 2000, height = 5500)\n",
    "print(h)\n",
    "\n",
    "\n",
    "### PROGENy ###\n",
    "library(progeny)\n",
    "# annotate matrix with HGNC symbols\n",
    "matched = match(rownames(mat), genes$ens_gene)\n",
    "rownames(mat) = genes$ext_gene[matched]\n",
    "\n",
    "pathways <- progeny(mat, \n",
    "                    scale=T,\n",
    "                    organism=\"Human\")\n",
    "pathways_tidy <- as.data.frame(pathways)\n",
    "pathways_tidy$sample <- rownames(pathways)\n",
    "pathways_tidy$Treatment <- c(\"Sensitive\", \"Sensitive\", \"Sensitive\",\n",
    "                             \"Resistant\", \"Resistant\", \"Resistant\")\n",
    "pathways_tidy <- gather(pathways_tidy, key = \"Pathway\", value = \"Activity\",\n",
    "                        -sample, -Treatment)\n",
    "\n",
    "testPathway <- function(pathway){\n",
    "  pathways_filtered <- filter(pathways_tidy, Pathway == pathway)\n",
    "  model_result <- summary(lm(Activity ~ Treatment, data=pathways_filtered))\n",
    "  results <- data.frame(pathway = pathway,\n",
    "                        Coefficient = model_result$coefficients[2,1],\n",
    "                        pval = model_result$coefficients[2,4])\n",
    "  return(results)\n",
    "}\n",
    "\n",
    "pathway_test <- lapply(colnames(pathways), testPathway)\n",
    "pathway_test <- do.call(\"rbind\", pathway_test)\n",
    "pathway_test$padj <- p.adjust(pathway_test$pval, method = \"BH\")\n",
    "\n",
    "sig_pathways <- filter(pathway_test, padj <= 0.05) %>%\n",
    "  pull(pathway)\n",
    "\n",
    "pheatmap(t(pathways),\n",
    "         colorRampPalette(c(\"black\", \"white\", \"#941A2B\"))(100),\n",
    "         breaks = seq(-2, 2, length.out=101),\n",
    "         cluster_cols=F,\n",
    "         clustering_method = \"ward.D2\",\n",
    "         border_color = \"black\",\n",
    "         filename=\"~/RNAseq/results/rlog_progeny_rr.png\",\n",
    "         width=3.2, height=4.2)\n",
    "\n",
    "### Overlap of DEGs associated with Cisplatin Resistance, Cancer Stemness, and EMT\n",
    "#Cisplatin Resistance (Huang D, 2021)\n",
    "CR <- which(df$symbol %in% Platinum_Resistent_Genes$`HUGO Gene symbol`)\n",
    "df$Cisplatin_Resistance <- \"\"\n",
    "df$Cisplatin_Resistance[CR] <- df$symbol[CR]\n",
    "\n",
    "CSC <- which(df$symbol %in% BCSCdb$...1)\n",
    "df$CSC <- \"\"\n",
    "df$CSC[CSC] <- df$symbol[CSC]\n",
    "\n",
    "EMT <- which(df$symbol %in% EMT_gene_sig$Gene)\n",
    "df$EMT <- \"\"\n",
    "df$EMT[EMT] <- df$symbol[EMT]\n",
    "\n",
    "write.table(df, file.path(\"~/RNAseq/results/20240107_RR_CSC_EMT_RES_markers.txt\"), row.names=TRUE, sep=\"\\t\",\n",
    "            quote=FALSE)\n",
    "\n",
    "### Volcano Plot ###\n",
    "\n",
    "labels_volcano <- c(\"CAV1\", \"WNT5A\", \"POSTN\", \"ITGA6\", \"CCN2\", \n",
    "                    \"COL1A1\", \"SNAI2\", \"SLC7A5\", \"ERBB3\", \"MMP10\", \n",
    "                    \"LUM\", \"COL6A3\", \"SPP1\", \"COL16A1\", \"FGF9\",\n",
    "                    \"PLCB\", \"POU5F1B\", \"BGN\", \"SOX8\", \"ID1\", \"KRT8\")\n",
    "\n",
    "label_rows_rr <- which(df$symbol %in% labels_volcano)\n",
    "df$labels <- \"\" #create column for volcano plot genes\n",
    "df$labels[label_rows_rr] <- df$symbol[label_rows_rr] #add gene names to label rows\n",
    "\n",
    "#labeling sig data\n",
    "df$sig <- \"No\"\n",
    "df$sig[which(df$padj <= 0.1 & abs(df$log2FoldChange) > 0.58)] <- \"Yes\"\n",
    "\n",
    "library(ggrepel)\n",
    "volcano_rr <- ggplot(df, aes(x=log2FoldChange, y=-log10(padj))) +\n",
    "  geom_point(size=2, alpha=0.5, aes(color=sig)) +\n",
    "  geom_vline(xintercept=c(-0.58, 0.58), linetype=2) +\n",
    "  geom_hline(yintercept=-log10(0.1), linetype=2) +\n",
    "  geom_text_repel(aes(label=labels),\n",
    "                  direction = \"both\",\n",
    "                  max.overlaps=Inf,\n",
    "                  size=4,\n",
    "                  segment.alpha=0.75,\n",
    "                  segment.color=\"grey50\",\n",
    "                  min.segment.length = 0.1,\n",
    "                  force=25,\n",
    "                  max.iter=5000) +\n",
    "  xlab(\"logFC\") + ylab(\"-log10(Adj. P-value)\") +\n",
    "  scale_color_manual(values=c(\"grey40\", \"#941A2B\")) +\n",
    "  theme_classic() +\n",
    "  xlim(c(-10,10)) + ylim(c(0,500)) +\n",
    "  theme(legend.position=\"none\",\n",
    "        axis.text=element_text(size=12, color=\"black\"),\n",
    "        axis.title=element_text(size=14, color=\"black\"))\n",
    "\n",
    "ggsave(volcano_rr, file=\"~/RNAseq/results/20240107_rr_volcano_plot_labels.pdf\", width = 7, height = 7)\n",
    "ggsave(volcano_rr, file=\"~/RNAseq/results/20240107_rr_volcano_plot_labels.png\", width = 7, height = 7, dpi = 300)\n",
    "\n",
    "\n",
    "### Box and Whiskers Plots for Individual Genes ###\n",
    "#rename table columns to make the names uniform (cell line_treatment_replicate number)\n",
    "gene_exp <- assay(rlog(dds_1, blind = FALSE))\n",
    "gene_df <- as.data.frame(gene_expr)\n",
    "gene_df$SYMBOL <- rownames(gene_df)\n",
    "gene_df <- gene_df %>% \n",
    "  rename(\"PDX4_SE_1\" = \"PDX4_SE1\",\n",
    "         \"PDX4_SE_2\" = \"PDX4_SE2\",\n",
    "         \"PDX4_SE_3\" = \"PDX4_SE3\",\n",
    "         \"PDX4_CR_1\" = \"PDX4_CR1\",\n",
    "         \"PDX4_CR_2\" = \"PDX4_CR2\",\n",
    "         \"PDX4_CR_3\" = \"PDX4_CR3\",)\n",
    "\n",
    "exp_tidy_rr <- gather(gene_df, \"Sample\", \"Expression\", -SYMBOL)\n",
    "exp_tidy_rr <- separate(exp_tidy_rr, Sample, into=c(\"Cell_Line\", \"Treatment\", \"Replicate\"))\n",
    "exp_tidy_rr$Treatment <- factor(exp_tidy_rr$Treatment, levels = c(\"SE\", \"CR\"))\n",
    "\n",
    "emt_csc_markers <- c(\"CDH2\", \"VIM\", \"TGFB1\", \"TGFB2\", \"ZEB2\",\n",
    "                     \"BMI1\", \"ENG\", \"MYC\", \"NOTCH1\", \"POU5F1B\")\n",
    "\n",
    "rr_exp <- gene_df %>% filter(SYMBOL %in% emt_csc_markers)\n",
    "rr_exp <- rr_exp[,c(\"SYMBOL\", \n",
    "                    \"PDX4_SE_1\", \"PDX4_SE_2\", \"PDX4_SE_3\",\n",
    "                    \"PDX4_CR_1\", \"PDX4_CR_2\", \"PDX4_CR_3\")]  \n",
    "rr_exp <- tidyr::gather(rr_exp, key=\"Sample\", value=\"Expression\", \n",
    "                        -SYMBOL)\n",
    "rr_exp$Condition <- ifelse(rr_exp$Sample %in% c(\"PDX4_SE_1\",\n",
    "                                                \"PDX4_SE_2\",\n",
    "                                                \"PDX4_SE_3\"),\n",
    "                           \"Sensitive\", \"Resistant\")\n",
    "\n",
    "rr_exp$Condition <- factor(rr_exp$Condition, levels = c(\"Sensitive\", \"Resistant\"))\n",
    "rr_exp$SYMBOL <- factor(rr_exp$SYMBOL, levels = c(\"CDH2\", \"BMI1\", \"VIM\", \"ENG\", \"TGFB1\", \"MYC\",\n",
    "                                                      \"TGFB2\", \"NOTCH1\", \"ZEB2\", \"POU5F1B\"))\n",
    "\n",
    "exp_plot <- ggplot(rr_exp, aes(x=Condition, y=Expression)) +\n",
    "  geom_boxplot(color=\"black\", aes(fill=Condition)) +\n",
    "  geom_point(color=\"black\", size=2.5) +\n",
    "  scale_fill_manual(values=c(\"gray25\", \"#941A2B\")) +\n",
    "  ylab(\"Normalized Counts\") + xlab(\"\") +\n",
    "  facet_wrap(~SYMBOL, scales=\"free\", ncol = 2 ) +\n",
    "  theme_bw() +\n",
    "  theme(legend.position=\"none\",\n",
    "        axis.text.y=element_text(size=12, color=\"black\"),\n",
    "        axis.text.x=element_text(size=12, color=\"black\", angle=45, hjust=1),\n",
    "        axis.title = element_text(size=12),\n",
    "        strip.background = element_blank(),\n",
    "        strip.text = element_text(size=14),\n",
    "        panel.grid=element_blank())\n",
    "exp_plot\n",
    "\n",
    "ggsave(exp_plot, filename=\"~/RNAseq/results/20240108_single_genes.png\", width=3, height=10)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
